### PR TITLE
Support/use multi-level caching that can share cache between instances

### PR DESCRIPTION
**Pull request by Page-**
Change-type: patch

### CODE REVIEW

File: src/infra/cache/index.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		cacheKey?: string;
+		promise: true;
+		primitive: true;
+		preFetch?: true | number;
+		maxAge: number;
+		max?: number;
+	},
+) => {
+	const {
+		cacheKey = fn.name,
```
Comment (thgreasi) [inline src/infra/cache/index.ts:23]: How about enforcing a `featureName.fnName` pattern?

**Review by thgreasi — COMMENTED** [review]
This looks sooo great.
Wonder how much it will decrease our S3 bill.

File: src/infra/cache/index.ts (hunk 1 above)
Comment (Page-) [reply src/infra/cache/index.ts:24]: I don't think it's possible to enforce as there's no way of knowing the feature name, beyond that this is "infra" code so shouldn't really have any specific knowledge and ideally will get moved out of open-balena-api altogether as it should be generically useful

**Comment by Page-** [comment]
@balena-ci I self-certify!

**Pull request merged**